openmls/src/treesync/node/leaf_node.rs

```diff
@@ -545,123 +571,200 @@ impl LeafNode {
                 }
             }
+
+            // ... and credential types MUST be listed in the LeafNode's capabilities field.
+            // TODO(#1200)
         }
-        Ok(())
+
+        Ok(self)
     }
 
-    /// Check whether the this leaf node supports all the required extensions
-    /// in the provided list.
-    #[cfg(test)]
-    pub(crate) fn check_extension_support(
+    /// Verify that the extensions in the LeafNode are supported.
+    /// * Check that the ID for each extension in the extensions field is listed in the capabilities.extensions field of the LeafNode.
+    pub(crate) fn validate_that_capabilities_contain_extension_types(
         &self,
-        extensions: &[ExtensionType],
-    ) -> Result<(), TreeSyncError> {
-        for required in extensions.iter() {
-            if !self.supports_extension(required) {
-                return Err(TreeSyncError::UnsupportedExtension);
+    ) -> Result<&Self, LeafNodeValidationError> {
+        for id in self
+            .payload
+            .extensions
+            .iter()
+            .map(Extension::extension_type)
+        {
+            if !self.supports_extension(&id) {
+                return Err(LeafNodeValidationError::ExtensionsNotInCapabilities);
             }
         }
-        Ok(())
+
+        Ok(self)
     }
 
-    /// Expose [`new_with_key`] for tests.
-    #[cfg(any(feature = "test-utils", test))]
-    pub(crate) fn create_new_with_key(
-        encryption_key: HpkePublicKey,
-        credential_bundle: &CredentialBundle,
-        leaf_node_source: LeafNodeSource,
-        capabilities: Capabilities,
-        extensions: Extensions,
-        backend: &impl OpenMlsCryptoProvider,
-    ) -> Result<Self, LibraryError> {
-        Self::new_with_key(
-            encryption_key,
-            credential_bundle,
-            leaf_node_source,
-            capabilities,
-            extensions,
-            backend,
-        )
+    /// Check that the credential type used in the leaf node is included in the credentials.
+    pub(crate) fn validate_that_capabilities_contain_credential_type(
+        &self,
+    ) -> Result<&Self, LeafNodeValidationError> {
+        if !self
+            .payload
+            .capabilities
+            .credentials
+            .contains(&self.payload.credential.credential_type())
+        {
+            return Err(LeafNodeValidationError::CredentialNotInCapabilities);
+        }
+
+        Ok(self)
     }
 
-    /// Returns the [`Lifetime`] if present.
-    /// `None` otherwise.
-    pub(crate) fn life_time(&self) -> Option<&Lifetime> {
-        if let LeafNodeSource::KeyPackage(life_time) = &self.payload.leaf_node_source {
-            Some(life_time)
-        } else {
-            None
+    /// Validate that the signature key is unique among the members of the group.
+    #[allow(unused)]
+    pub(crate) fn validate_that_signature_key_is_unique(
+        &self,
+        signature_keys: &[SignaturePublicKey],
```

Comment:
It might make sense to take an iterator over `Member` here. But we won't know for sure until we use it, I suppose.

Reply:
I can try it out in a follow-up when we introduce the type safe leaf node validation. Currently, not every validation step is executed because I was unsure where to do it. Thus, it is difficult to see what works and what not.